projects/driver_controls/test/test_ads1015.c

```diff
@@ -0,0 +1,97 @@
+  uint8_t *callback_called = context;
+  (*callback_called) = true;
+}
+
+void setup_test(void) {
+  gpio_init();
+  interrupt_init();
+  gpio_it_init();
+  I2CSettings i2c_settings = {
+    .speed = I2C_SPEED_FAST, .scl = { GPIO_PORT_B, 10 }, .sda = { GPIO_PORT_B, 11 }
+  };
+  i2c_init(I2C_PORT_2, &i2c_settings);
+}
+
+void teardown_test(void) {}
+
+void test_ads1015_init_invalid_input(void) {
+  GPIOAddress ready_pin = { GPIO_PORT_B, 2 };
+  Ads1015Storage storage;
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK,
+                    ads1015_init(&storage, I2C_PORT_2, ADS1015_ADDRESS_GND, &ready_pin));
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS,
+                    ads1015_init(NULL, I2C_PORT_2, ADS1015_ADDRESS_GND, &ready_pin));
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS,
+                    ads1015_init(&storage, I2C_PORT_2, ADS1015_ADDRESS_GND, NULL));
+}
+
+void test_ads_config_channel_invalid_input(void) {
+  Ads1015Storage storage;
+  GPIOAddress ready_pin = { GPIO_PORT_B, 2 };
+
+  ads1015_init(&storage, I2C_PORT_2, ADS1015_ADDRESS_GND, &ready_pin);
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK, ads1015_configure_channel(&storage, ADS1015_CHANNEL_0, true,
+                                                              prv_callback, &storage));
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK, ads1015_configure_channel(&storage, ADS1015_CHANNEL_0, false,
+                                                              prv_callback, &storage));
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK,
+                    ads1015_configure_channel(&storage, ADS1015_CHANNEL_0, true, NULL, &storage));
+  TEST_ASSERT_EQUAL(
+      STATUS_CODE_INVALID_ARGS,
+      ads1015_configure_channel(&storage, NUM_ADS1015_CHANNELS, true, prv_callback, &storage));
+  TEST_ASSERT_EQUAL(
+      STATUS_CODE_INVALID_ARGS,
+      ads1015_configure_channel(NULL, ADS1015_CHANNEL_1, true, prv_callback, &storage));
+}
+
+void test_ads1015_read_invalid_input(void) {
+  int16_t reading;
+  Ads1015Storage storage;
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK, ads1015_read_converted(&storage, ADS1015_CHANNEL_0, &reading));
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK, ads1015_read_raw(&storage, ADS1015_CHANNEL_2, &reading));
+
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS,
+                    ads1015_read_converted(NULL, ADS1015_CHANNEL_2, &reading));
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS,
+                    ads1015_read_converted(&storage, ADS1015_CHANNEL_2, NULL));
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS,
+                    ads1015_read_converted(&storage, NUM_ADS1015_CHANNELS, &reading));
+
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, ads1015_read_raw(NULL, ADS1015_CHANNEL_2, &reading));
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, ads1015_read_raw(&storage, ADS1015_CHANNEL_2, NULL));
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS,
+                    ads1015_read_raw(&storage, NUM_ADS1015_CHANNELS, &reading));
+}
+
+void test_ads1015_channel_callback(void) {
```

Comment:
Can you add more test cases for your different scenarios? You have a lot of cases like

* No channels enabled
* Going from channel(s) enabled to disabled
* Enabling a new channel during another channel's conversion

etc.

Basically, it's important to test the edge cases. All the manual testing you've done is great, but unit tests are much easier to use.